feat(boilerplate): ES6, New/Remove User Stories, Add Tests, Fix error in assertion analyser

SaintPeter opened the pull request:
I have made a number of changes to this project:
1. Updated some of the libraries and moved code to `let`/`const`
2. Fix a bug with `assertion-analyser` which broke when there were "empty" tests
3. Improved clarity in user stories about the `title` field (Ref: https://github.com/freeCodeCamp/freeCodeCamp/issues/39630)
4. Added additional tests for the `delete` verb and one for the `post` comments, in line with item 3.
5. Removed InfoSec stories (Ref: https://github.com/freeCodeCamp/freeCodeCamp/issues/39435) 

## Testing
I have tested this locally and everything seems dandy.

## Additional testing needed
@scissorsneedfoodtoo - Can you please check against Repl.it?  @nhcarrigan was able to test this locally but ran into trouble getting it working on the Repl.it.
 
## Sync with /learn
I will be submitting a PR for the `learn` project once I complete this on my own and write the tests.

SaintPeter committed:
Use let/const defintions

SaintPeter committed:
Update package revisions
Add dotenv
Remove helmet

SaintPeter committed:
Add DELETE and bad id tests, Use let/const

SaintPeter committed:
Remove InfoSoc user stories, add missing title error definition and clarifications

SaintPeter committed:
Fix support for empty functions

SaintPeter committed:
Switch to let/const

naomi-lgbt commented:
To clarify the issue I was having:

I forked the example project on repl.it to have a standard functional code against which to test these changes. Because the dependencies were updated, I deleted the `package-lock.json` and ran `npm install` to get a clean set of packages. I added my `DB_URI` value to the `.env` file on my repl.it instance. I verified the connection by launching the page and using the HTML form to add a book to the library. The data was successfully written to my MongoDB cluster. I called the test runner with `npx mocha -u tdd tets/2_functional-tests.js`. The tests errored out with the message: 
```bash
Uncaught AssertionError: database error: expected [MongoParseError: URI malformed, cannot be parsed] to not exist
```

I could not get the tests to successfully run on repl.it, so I downloaded the project files from my instance as a `.zip`,  unpacked them, and ran the tests locally with the same process as above and encountered no error. The tests functioned as intended on my local instance.

SaintPeter committed:
Add `comment` field requirement and test

SaintPeter commented:
I added one more change, a failure and test for the `comment` field when POSTing to `/api/books/{_id}`.

I also added some minor styling to the `<code>` tags so they'll stand out a bit more.

naomi-lgbt commented:
Hey @SaintPeter - can we remove the [relevant line](https://github.com/freeCodeCamp/boilerplate-project-library/blame/gomix/README.md#L10) from the `README.md` as well?

SaintPeter committed:
Remove security features reference

SaintPeter commented:
> Hey @SaintPeter - can we remove the relevant line from the README.md as well?
Done.

scissorsneedfoodtoo commented:
@SaintPeter, thanks for all of your hard work here. Sure, I'll check out the changes you made over on Repl.it and get back to you.

We'll have some guidelines soon on how to handle these boilerplates as far as the README and user stories go, so please hang tight in the meantime.

SaintPeter commented:
> We'll have some guidelines soon on how to handle these boilerplates as far as the README and user stories go, so please hang tight in the meantime.

If you'd like me to move the User Stories to the README or remove them entirely and create a PR for the site, I can do that as well.

ShaunSHamilton commented:
@SaintPeter Rex, could you include the `package-lock.json` file in this PR, please? It could help https://github.com/freeCodeCamp/freeCodeCamp/issues/39731

SaintPeter commented:
@Sky020 Done.

SaintPeter committed:
Move to ES6 Formatting

SaintPeter committed:
Remove User Stories

SaintPeter commented:
Removed User Stories from HTML and README as per https://github.com/freeCodeCamp/freeCodeCamp/issues/39721

ShaunSHamilton commented on README.md:
```diff
@@ -1,12 +1,6 @@
-**FreeCodeCamp**- Information Security and Quality Assurance
-------
+# **freeCodeCamp** 
+### [Quality Assurance Projects - Personal Library](https://www.freecodecamp.org/learn/quality-assurance/quality-assurance-projects/personal-library)
 
```
```suggestion
# [Quality Assurance Projects - Personal Library](https://www.freecodecamp.org/learn/quality-assurance/quality-assurance-projects/personal-library)
```

ShaunSHamilton reviewed (changes requested):
Sorry if I was not clear with this:
- There should only be **1** line in the README. **No** _"freeCodeCamp"_

SaintPeter committed:
Update README.md

Co-authored-by: Shaun Hamilton <51722130+Sky020@users.noreply.github.com>

SaintPeter committed:
Remove Quality Assurance Prefix

scissorsneedfoodtoo commented on views/index.html:
```diff
@@ -1,85 +1,72 @@
-    <hr style='margin: 50px'>
+<head>
+  <title>Quality Assurance Project - Personal Library</title>
+  <meta name="description" content="A cool thing made with HyperDev">
+  <link id="favicon" rel="icon" href="https://hyperdev.com/favicon-app.ico" type="image/x-icon">
```
The meta description can be removed and we can use the indigo fCC favicon served from our CDN:

`<link rel="icon" type="image/png" href="https://cdn.freecodecamp.org/universal/favicons/favicon-16x16.png" />`

scissorsneedfoodtoo reviewed (changes requested):
Noticed one minor thing on the landing page. Also, the repository key can be updated to point to the boilerplate repo or removed entirely.

Those should be the only mentions of Hyperdev/Glitch left.

It seems like this boilerplate is missing a `sample.env ` and `.gitignore` file. @RandellDawson and @moT01, will we be providing both of those in the boilerplates? Or just the `.gitignore` file?

scissorsneedfoodtoo commented:
@SaintPeter, sorry, I forgot to write back about this earlier.

I forked the current boilerplate from the project page and mirrored all your changes there. The only problem I ran into was that Repl.it didn't pick up the `.replit` file the first time, and assumed the project was a Bash script. Once I added that in everything seemed to work fine. And when I forked the project a few more times it picked up all the files and would run.

SaintPeter committed:
Formatting Cleanup

SaintPeter commented:
> Also, the repository key can be updated to point to the boilerplate repo or removed entirely.

@scissorsneedfoodtoo I'm not sure what you mean by a "repository key"

I have fixed the other items, except I need the sample files . . . or you could just merge this thing and add the files yourselves.

scissorsneedfoodtoo commented:
@SaintPeter, sorry, I should have been more clear. I mean the repository key in the `package.json` file which still points to hyperdev. I'm not able to select that line in the file and request changes directly.

For the other files, sure, I can open up another PR to add those in. We've been going with pretty simple `.gitignore` and `sample.env` files in the demo projects repo.

scissorsneedfoodtoo committed:
feat: add sample.env and .gitignore files

scissorsneedfoodtoo committed:
fix: change DB_URI to DB in sample.env

SaintPeter committed:
Update repository key

SaintPeter commented:
@scissorsneedfoodtoo I have update the repo key as requested.

SaintPeter committed:
Add .gitignore and sample.env

SaintPeter commented:
Since I had it open, I removed the `README.md` and added the `sample.env` and `.gitignore`

RandellDawson commented:
Why was the README.md removed?  The only place the README.md file is supposed to be removed is in the demo-projects repo for the specific project.  For the project boilerplates, the README.md will just have a level 1 heading linking to project on `/learn`.

A side question is what is the actual branch that should be targeted?  This PR targets `gomix` vs. `master`.  I believe the repl.it links on `/learn` only work for what is on the `master` branch and not other branches.

SaintPeter commented:
Ok, oops.  I'll put the file back.

The default branch is `gomix` - I didn't choose it.  I don't know about the links, but you can set the default branch for a repo in the settings for the repo.  I'm pretty sure that whatever the default branch is is what gets linked to.

SaintPeter commented:
I have resolved all of the merge conflicts by rebasing against the current head.  This one should be ready for merge.

ShaunSHamilton commented on public/client.js:
```diff
@@ -1,9 +1,9 @@
   
   $.getJSON('/api/books', function(data) {
-    //var items = [];
+    //let  items = [];
```
Any particular reason this is here?

SaintPeter replied:
It was in the original boilerplate.  No idea why.  As you can tell, I did a search-and-replace on `var`, so this was picked up.  I didn't otherwise touch the client-side JS code.

I can remove it, of course.

ShaunSHamilton replied:
Oh, right. I missed that. It is not major, but if it serves no purpose, then it can only confuse.

SaintPeter replied:
Removed it.

scissorsneedfoodtoo commented:
@SaintPeter, awesome, thank you for going through and adding those things.

@RandellDawson, Repl.it should pick up on whatever the default branch is, so `gomix` should be fine here. We'll need help from Mrugesh to change the default branch from `gomix` to `main`, which I believe is what GitHub defaults to now.

SaintPeter commented:
You can switch the default branch of any repository to whatever you'd like:
![image](https://user-images.githubusercontent.com/553494/96447521-de8e5a00-11c6-11eb-9a01-777686d554e9.png)

You can also rename an existing branch:
https://jarv.is/notes/github-rename-master/
(I propose that we do this for the current `gomix` branch)

More details behind GitHub's "main" policy:
https://github.com/github/renaming

ShaunSHamilton reviewed (approved):
This LGTM 👍

scissorsneedfoodtoo commented:
Thanks for the detailed instructions @SaintPeter. I don't have the permissions to do that for any of the boilerplate repos, but we'll definitely switch them over as soon as possible.

Took another look and everything LGTM :+1: :+1:

RandellDawson commented on routes/api.js:
```diff
@@ -8,9 +8,9 @@
-var MongoClient = require('mongodb').MongoClient;
-var ObjectId = require('mongodb').ObjectId;
+const expect = require('chai').expect;
+const MongoClient = require('mongodb').MongoClient;
+const ObjectId = require('mongodb').ObjectId;
 const MONGODB_CONNECTION_STRING = process.env.DB;
 //Example connection: MongoClient.connect(MONGODB_CONNECTION_STRING, function(err, db) {});
```
I believe these lines can be removed from the boilerplate as users can write their own code to satisfy the route requirements.

naomi-lgbt replied:
Agreed.

RandellDawson reviewed (changes requested):
@SaintPeter See the following change suggested to complete this current PR.  You can see me [comment on the learn PR](https://github.com/freeCodeCamp/freeCodeCamp/pull/39642#pullrequestreview-528434305) for what an additional PR could do to satisfy the requirements of issue https://github.com/freeCodeCamp/freeCodeCamp/issues/39721.

Delete the following files:
- '/tests/1_unit-tests.js`
- `.bashrc`
- `.hyperdev-assets`
- `.profile`

SaintPeter committed:
Remove unneeded files and lines

SaintPeter committed:
Merge branch 'gomix' into feat/improve-user-stories-tests

SaintPeter commented:
Should be good to go.

moT01 reviewed (approved):
LGTM 🎉 Should be good to go once the learn PR hits production 😄

RandellDawson commented:
The /learn changes have hit production, so I am merging this now.
